File: java/05_047.java
import java.util.*;
import java.util.stream.*;
import java.io.*;
import java.math.*;

public class Main {
    public static void main(String[] args) {
        // Main method code goes here
    }
    
    public class Node {
        public int val;
        public Node left;
        public Node right;
        
        public Node(int val) {
            this.val = val;
        }
    }
    
    public List<Integer> inorderIterative(Node root) {
        List<Integer> result = new ArrayList<>();
        Stack<Node> stack = new Stack<>();
        Node current = root;
        
        while (current != null || !stack.isEmpty()) {
            while (current != null) {
                stack.push(current);
                current = current.left;
            }
            
            current = stack.pop();
            result.add(current.val);
            current = current.right;
        }
        
        return result;
    }
}

File: data structures and algorithms/cs/05_047.cs
using System;
using System.Collections.Generic;

public class MainClass {
    public static void Main(string[] args) {
        // Main method code goes here
    }
    
    public class Node {
        public int val;
        public Node left;
        public Node right;
        
        public Node(int val) {
            this.val = val;
        }
    }
    
    public List<int> InorderIterative(Node root) {
        List<int> result = new List<int>();
        Stack<Node> stack = new Stack<Node>();
        Node current = root;
        
        while (current != null || stack.Count > 0) {
            while (current != null) {
                stack.Push(current);
                current = current.left;
            }
            
            current = stack.Pop();
            result.Add(current.val);
            current = current.right;
        }
        
        return result;
    }

    public void PrintPreOrder(Node root) {
        if (root == null) {
            return;
        }
        Console.Write(root.val + " ");
        PrintPreOrder(root.left);
        PrintPreOrder(root.right);
    }
}
